### ISSUE TITLE
Wild idea: use prototypes to structurally share non modified properties

### ISSUE THREAD

**Issue by mweststrate** [immerjs/immer#501]
i have a immature idea
if we have a large obj , such as
```
var obj = {
a1:1,
a2:2,
a3:3,
...
a1000000:1000000
}
```
now, immer looks like it's going to copy the object all key again, like{...obj}, it's will slow
how about just like this:
var objProto = {\__proto__:obj}
then, we can access property like objProto.a1,
and we can modify property like objProto.a1 = 2,( and we Not modify origin data obj )

_Originally posted by @zhangenming in https://github.com/immerjs/immer/issues/462#issuecomment-554613735_

----

Thoughts: might safe time and space! 

Might mess up iteration since a lot of properties won't be "own". 

Could be an experimental opt in option?

**Comment by eddyw** [comment]
I tried something like that before. At first it may seem to work but you eventually end up with a very nested prototype chain:
```obj.__proto__.__proto__.__proto__```
This means a few things:
- Property lookup becomes slow
- No garbage collection since the new object will have a reference to its old version in prototype
- Only "safe" as long as the object in `__proto__` is frozen

An alternative way could be to force objects with > 32 property keys to become "Fast Properties" instead of letting them fall to "Slow/Dictionary based". Check here: https://github.com/v8/v8/blob/6.0.122/test/mjsunit/fast-prototype.js#L48-L63 
And a lib that already does this: https://github.com/sindresorhus/to-fast-properties/blob/master/index.js

The basic idea is that, every object that is in prototype is optimized  as "Fast Properties" once ICs (Inline Caches) find them. Thus, this line that doesn't seem to make any sense but in fact just makes ICs find them:
```
  function ic() { return typeof receiver.foo; }
  ic();
  ic();
```

So, maybe as a flag, option, or something, Immer could just do this:
```
DoProtoMagic(obj, objKeys.length > 32)
```
Why 32? Last time I checked, this is the max number of allocated fast properties an object can have before falling back to dictionary based. In any case, there is no much or not at all performance gain with these small objects but we could try if there is any gain in .. idk, 500 - 5000 props? 🙈 

Ref: https://v8.dev/blog/fast-properties
